### PR TITLE
[trainer] sanitize test case name and test identifier consistently when looking up test failures

### PR DESCRIPTION
**Pull request by TheMetalCode**
### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.
- [ ] I've added or updated relevant unit tests.

### Motivation and Context

I recently had an issue in CI builds where failing tests were not being reported as failures once trainer converted the xcresult to junit xml. After digging in and debugging, I found that the existing logic in `Trainer::XCResult::ActionTestMetadata#find_failure` was expecting the test identifier to have a particular format that didn't end up matching. This led me to the insight that instead of having this expectation of the identifier, we could sanitize both it and the test case name consistently, then compare them for a potential match.

Using xcresult artifacts from the failing CI runs, this fix was effective. Unfortunately, since those xcresult artifacts contain proprietary data, I've yet to figure out how to unit test this safely. 🙈

### CODE REVIEW

File: trainer/lib/trainer/xcresult.rb [hunk 1]
```diff
@@ -184,16 +184,12 @@ def find_failure(failures)
...
           found_failure = failures.find do |failure|
-            # Clean test_case_name to match identifier format
-            # Sanitize for Swift by replacing "." for "/"
-            # Sanitize for Objective-C by removing "-", "[", "]", and replacing " " for ?/
-            sanitized_test_case_name = failure.test_case_name
-                                              .tr(".", "/")
-                                              .tr("-", "")
-                                              .tr("[", "")
-                                              .tr("]", "")
-                                              .tr(" ", "/")
```
Comment (TheMetalCode) [inline trainer/lib/trainer/xcresult.rb:195]: In my case, this ended up being the unmet expectation - not all spaces were replaced by `/` in the `identifier`. So the comparison ended up being [somewhat redacted due to proprietary info]:

```
[1] pry(#<Trainer::XCResult::ActionTestMetadata>)> self.identifier
=> "MyPrivateSpec/my test condition, should return an error()"
[2] pry(#<Trainer::XCResult::ActionTestMetadata>)> sanitized_test_case_name
=> "MyPrivateSpec/my/test/condition,/should/return/an/error()"
```

Anyways, the bet I'm making with this PR is: rather than have any expectation of how `self.identifier` is formatted, sanitize both it and `failure.test_case_name` in the same fashion then compare.

**Review by rogerluan — CHANGES_REQUESTED** [review]
Hi @TheMetalCode 👋 

Thanks for this PR! Your rationale makes sense to me, and I think this is a good solution moving forward.

Do you mind adding unit tests to this PR, hopefully with real-life strings for both test case name and identifiers? Ideally we should cover the happy path (scenarios where the previous implementation used to work) as well as the edge case you were seeing (so that it fails, when run using the previous comparison algorithm, but succeeds now using the new algorithm). That would be super!

Let me know if you need any support adding these tests 🤗 thanks again for your hard work! 🙏 

PS: If you could also add steps to reproduce this issue locally so we can verify your results, that would be great. The simpler these steps are, the better (extra points if you can provide a demo project where the issue is reproducible) 🙇

**Comment by rogerluan** [comment]
Heads up: no need to implement the unit tests in this PR. I decided to merge part of this PR's solution into the PR https://github.com/fastlane/fastlane/pull/21432 where we already have unit tests and a sample project, so once we merge that PR, this one can be closed 🙇 

Thanks for the contribution in this PR! 🙌

**Comment by TheMetalCode** [comment]
> Heads up: no need to implement the unit tests in this PR. I decided to merge part of this PR's solution into the PR #21432 where we already have unit tests and a sample project, so once we merge that PR, this one can be closed 🙇
> 
> Thanks for the contribution in this PR! 🙌

That's great, thank you!

My struggle with unit tests on this one is that I had them locally but they were relying on sample xcresult bundles from a private project and I didn't know of a way to edit xcresult files to make them non-proprietary. This private project is using Quick/Nimble though, didn't realize that in and of itself would create the error condition.

Anyways, happy to have influenced #21432, looking forward to the fix!
